src/db.c

```diff
@@ -1914,98 +1913,93 @@ int xreadGetKeys(struct redisCommand *cmd, robj **argv, int argc, getKeysResult
     return num;
 }
 
+/* In cluster mode, links to the next and previous entries in the same slot are
+ * stored in the dict entry metadata. */
+#define dictEntryNextInSlot(de) (((dictEntry **)dictMetadata(de))[0])
+#define dictEntryPrevInSlot(de) (((dictEntry **)dictMetadata(de))[1])
+
 /* Slot to Key API. This is used by Redis Cluster in order to obtain in
  * a fast way a key that belongs to a specified hash slot. This is useful
  * while rehashing the cluster and in other conditions when we need to
  * understand if we have keys for a given hash slot. */
-void slotToKeyUpdateKey(sds key, int add) {
+void slotToKeyUpdateKey(dictEntry *entry, int add) {
```

Comment:
This should be refactored into 2 separate functions.  Indications:
* Almost the entire code is different, based on a boolean flag.
* In every place where this is called, the boolean flag is hard-coded.

Reply:
Makes sense, they do have less in common then they used to.

Reply:
Will do. It's only called in slotToKeyAdd and slotToKeyDel, so I'll put the implementation in each of these.